services/optimization.js

```diff
@@ -0,0 +1,162 @@
+          bearing: v.arrayOf(v.range([0, 360])),
+          radius: v.oneOfType(v.number, v.equal('unlimited'))
+        })
+      )
+    ),
+    annotations: v.arrayOf(v.oneOf('duration', 'distance', 'speed')),
+    geometries: v.string,
+    language: v.string,
+    overview: v.oneOf('simplified', 'full', 'false'),
+    roundtrip: v.boolean,
+    steps: v.boolean,
+    source: v.oneOf('any', 'first'),
+    destination: v.oneOf('any', 'last'),
+    distributions: v.arrayOf(
+      v.shape({
+        pickup: v.number,
+        dropoff: v.number
+      })
+    )
+  })(config);
+
+  config.profile = config.profile || 'driving';
+
+  var path = {
+    coordinates: [],
+    approach: [],
+    bearing: [],
+    radius: [],
+    distributions: []
+  };
+
+  var waypointCount = config.waypoints.length;
+  if (waypointCount < 2 || waypointCount > 12) {
+    // updated to reflect opt restrictions
+    throw new Error(
+      'waypoints must include between 2 and 12 OptimizationWaypoints'
+    );
+  }
+
+  /**
+   * @typedef {Object} OptimizationWaypoint
+   * @property {Coordinates} coordinates
+   * @property {'unrestricted'|'curb'} [approach="unrestricted"] - Used to indicate how requested routes consider from which side of the road to approach the waypoint.
+   * @property {[number, number]} [bearing] - Used to filter the road segment the waypoint will be placed on by direction and dictates the angle of approach.
+   *   This option should always be used in conjunction with a `radius`. The first value is an angle clockwise from true north between 0 and 360,
+   *   and the second is the range of degrees the angle can deviate by.
+   * @property {number|'unlimited'} [radius] - Maximum distance in meters that the coordinate is allowed to move when snapped to a nearby road segment.
+   */
+  config.waypoints.forEach(function(waypoint) {
+    path.coordinates.push(
+      waypoint.coordinates[0] + ',' + waypoint.coordinates[1]
+    );
+
+    // join props which come in pairs
+    ['bearing'].forEach(function(prop) {
+      if (waypoint.hasOwnProperty(prop) && waypoint[prop] != null) {
+        waypoint[prop] = waypoint[prop].join(',');
+      }
+    });
+
+    ['approach', 'bearing', 'radius'].forEach(function(prop) {
+      if (waypoint.hasOwnProperty(prop) && waypoint[prop] != null) {
+        path[prop].push(waypoint[prop]);
+      } else {
+        path[prop].push('');
+      }
+    });
+  });
+
+  ['approach', 'bearing', 'radius'].forEach(function(prop) {
+    // avoid sending params which are all `;`
+    if (
+      path[prop].every(function(char) {
+        return char === '';
+      })
+    ) {
+      delete path[prop];
+    } else {
+      path[prop] = path[prop].join(';');
+    }
+  });
+
+  if (config.distributions) {
+    var distributionsList = config.distributions.map(function(dist) {
+      return [dist.pickup, dist.dropoff];
+    });
+    path.distributions.push(distributionsList.join(';'));
```

Comment:
It's odd to me that `path.distributions` is still an array but will only ever end up with one item in it, right?

Reply:
Thanks for this! Per discussion: 

1. Moved the `join` piece to the `forEach` loop along with `approach`, `bearing`, etc. 
2. Pushed each pickup/dropoff pair to `path.distributions` instead of putting them into various unnecessary arrays first.